dice-engine/test/dice_engine/dice_test.clj

```diff
@@ -2,182 +2,221 @@
-                             {:value           3
-                              :discarded       true
-                              :faces           4
-                              :previous-values []}
-                             {:value           4
-                              :discarded       false
-                              :faces           4
-                              :previous-values []}]})))
-  (testing "should return mark all dice as discarded and update :numeric-value to 0 if none match"
-    (is (= dice/keep {:numeric-value 8
-                      :dice          [
-                                      {:value           1
-                                       :discarded       false
-                                       :faces           4
-                                       :previous-values []}
-                                      {:value           3
-                                       :discarded       false
-                                       :faces           4
-                                       :previous-values []}
-                                      {:value           4
-                                       :discarded       false
-                                       :faces           4
-                                       :previous-values []}]}
-           2)
-        '{:numeric-value 0
-          :dice          [
-                          {:value           1
-                           :discarded       true
-                           :faces           4
-                           :previous-values []}
-                          {:value           3
-                           :discarded       true
-                           :faces           4
-                           :previous-values []}
-                          {:value           4
-                           :discarded       true
-                           :faces           4
-                           :previous-values []}]})))
+  (testing "should return dice that match the value"
+    (is (=
+          (dice/keep dice/match 4 dice-1)
+          [{:id              2
+            :value           4
+            :faces           6
+            :previous-values []}])))
+
+  (testing "should return empty vector if none of the dice values match"
+    (is (=
+          (dice/keep dice/match 2 dice-1)
+          []))))
+
+(deftest reroll-matched-test
+  (testing "should reroll the matched die"
+    (with-redefs-fn {#'dice/rand-int-natural (fn [_] 2)}
+      #(is (=
+             (dice/reroll-matched dice/match 3 dice-1)
+             [{:id              1
+               :value           2
+               :faces           6
+               :previous-values [3]}
+              {:id              2
+               :value           4
+               :faces           6
+               :previous-values []}]))))
+  (testing "should reroll until no die qualifies selector"
+    (let [rand-ints [5 9 1]]
```

Comment:
`with-predictable-die-rolls` would be a good candidate for a macro to help with tests for when you want to write some macros.